src/main/java/com/mongodb/spark/sql/connector/config/MongoConfig.java

```diff
@@ -17,191 +17,198 @@
-    INPUT,
-    OUTPUT
-  }
+  String CLIENT_FACTORY_DEFAULT = DefaultMongoClientFactory.class.getName();
 
   /**
-   * Creates a new {@code MongoConfig} for inputting data.
+   * The connection string configuration key
    *
-   * @param options the user provided options
-   * @return the MongoConfig.
+   * <p>{@value}
    */
-  public static MongoConfig createInputConfig(final CaseInsensitiveStringMap options) {
-    return new MongoConfig(options, UsageMode.INPUT);
-  }
-
+  String CONNECTION_STRING_CONFIG = "connection.uri";
   /**
-   * Creates a new {@code MongoConfig} for outputting data.
+   * The default connection string configuration value
    *
-   * @param options the user provided options
-   * @return the MongoConfig.
+   * <p>{@value}
    */
-  public static MongoConfig createOutputConfig(final CaseInsensitiveStringMap options) {
-    return new MongoConfig(options, UsageMode.OUTPUT);
-  }
+  String CONNECTION_STRING_DEFAULT = "mongodb://localhost:27017/";
 
-  private final CaseInsensitiveStringMap originals;
-  private final CaseInsensitiveStringMap values;
-  private final UsageMode usageMode;
-  private transient MongoClientFactory mongoClientFactory;
-
-  private MongoConfig(final CaseInsensitiveStringMap originals, final UsageMode usageMode) {
-    this.originals = originals;
-    this.usageMode = usageMode;
-
-    String ignorePrefix = usageMode == UsageMode.INPUT ? MONGO_OUTPUT_PREFIX : MONGO_INPUT_PREFIX;
-    String overridePrefix = usageMode == UsageMode.INPUT ? MONGO_INPUT_PREFIX : MONGO_OUTPUT_PREFIX;
-
-    Map<String, String> usageSpecificSettings = new HashMap<>();
-    List<String> defaults = new ArrayList<>();
-    List<String> overrides = new ArrayList<>();
-
-    originals.keySet().stream()
-        .filter(k -> k.startsWith(MONGO_PREFIX))
-        .forEach(
-            k -> {
-              if (k.startsWith(overridePrefix)) {
-                overrides.add(k);
-              } else if (!k.startsWith(ignorePrefix)) {
-                defaults.add(k);
-              }
-            });
-
-    defaults.forEach(
-        k -> usageSpecificSettings.put(k.substring(MONGO_PREFIX.length()), originals.get(k)));
-    overrides.forEach(
-        k -> usageSpecificSettings.put(k.substring(overridePrefix.length()), originals.get(k)));
-
-    this.values = new CaseInsensitiveStringMap(usageSpecificSettings);
-  }
+  /**
+   * The database name config
+   *
+   * <p>{@value}
+   */
+  String DATABASE_NAME_CONFIG = "database";
 
   /**
-   * Convert the configuration for use when inputting data
+   * The collection name config
    *
-   * @return the MongoConfig to use.
+   * <p>{@value}
    */
-  public MongoConfig toInputMongoConfig() {
-    if (this.usageMode == UsageMode.INPUT) {
-      return this;
-    }
-    return new MongoConfig(originals, UsageMode.INPUT);
-  }
+  String COLLECTION_NAME_CONFIG = "collection";
+
+  /** @return the options for this MongoConfig instance */
+  Map<String, String> getOptions();
```

Comment:
If `MongoConfig.getOriginals`/`getOptions` must return either a copy or an unmodifiable map, and `MongoConfig.withOptions` must not mutate the state of the `MongoConfig`, then we should document these methods as such and document `MongoConfig` as being unmodifiable<sup>1</sup>.

<sup>1</sup> We can't document it as immutable, because that implies both thread-safety (current implementations are not thread-safe) and robustness to publications via a data race (which is a different property, e.g., `AtomicInteger` is thread-safe, but one is allowed to observe corrupted state if accessing the object via a data race).

Reply:
I dont think we need to anything more here in the interface. The implementations do the right thing.